tests/test_pyrflx.py

```diff
@@ -674,12 +676,12 @@ def test_parsing_ethernet_2_vlan(self) -> None:
         self.assertEqual(int("8100", 16), self.frame.get("TPID"))
         self.assertEqual(int("1", 16), self.frame.get("TCI"))
 
-        k = self.frame._fields["Payload"].length
+        k = self.frame._fields["Payload"].typeval.size
         assert isinstance(k, Number)
         self.assertEqual(47, k.value // 8)
 
         self.assertTrue(self.frame.valid_message)
-        self.assertEqual(self.frame.value, msg_as_bytes)
+        self.assertEqual(self.frame.bytestring, msg_as_bytes)
 
     def test_ethernet_invalid_ethernet_2_too_short(self) -> None:
```

Comment:
Please make the naming of the tests consistent:

- `Test_Parsing_Invalid_Ethernet_II_Too_Short` in `Ethernet.Tests` =>
test_ethernet_parsing_invalid_ethernet_2_too_short **(currently: test_ethernet_invalid_ethernet_2_too_short)**
- `Test_Parsing_Incomplete` in `Ethernet.Tests` =>
test_ethernet_parsing_incomplete **(currently: test_parsing_incomplete)**

Please also separate the "converted tests" from the other unit tests, e.g., by putting all them at the end of the file. It would probably good to add comments to make the origin of these tests clearer. (I know you had them before and you probably removed them because of one of my comments. Sorry!)

Reply:
Fixed by a02992f

Reply:
They are still not correctly sorted. Here is the current list:

```
tests/test_pyrflx.py:    # rflx-ethernet-tests.adb
tests/test_pyrflx.py:    def test_ethernet_parsing_ethernet_2(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_parsing_ieee_802_3(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_parsing_ethernet_2_vlan(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_invalid_ethernet_2_too_short(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_invalid_ethernet_2_too_long(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_parsing_invalid_ethernet_2_undefined_type(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_ieee_802_3_invalid_length(self) -> None:
tests/test_pyrflx.py:    def test_ethenet_parsing_incomplete(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_generating_ethernet_2(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_generating_ieee_802_3(self) -> None:
tests/test_pyrflx.py:    def test_ethernet_generating_ethernet_2_vlan(self) -> None:
tests/test_pyrflx.py:    # def test_generating_ethernet_2_vlan_dynamic(self) -> None:
tests/test_pyrflx.py:    # def test_parsing_ipv4_in_ethernet(self) -> None:
tests/test_pyrflx.py:    # def test_generating_ipv4_in_ethernet(self) -> None:
tests/test_pyrflx.py:    # def test_parsing_udp_in_ipv4(self) -> None:
tests/test_pyrflx.py:    # def test_parsing_udp_in_ipv4_in_ethernet(self) -> None:
tests/test_pyrflx.py:    # rflx-in_ipv4-tests.adb
tests/test_pyrflx.py:    def test_ethernet_generating_udp_in_ipv4_in_ethernet(self) -> None:
tests/test_pyrflx.py:    # rflx-ipv4-tests.adb
tests/test_pyrflx.py:    def test_ipv4_parsing_ipv4(self) -> None:
tests/test_pyrflx.py:    # def test_parsing_ipv4_with_options(self) -> None:
tests/test_pyrflx.py:    # def test_parsing_ipv4_option(self) -> None:
tests/test_pyrflx.py:    def test_ipv4_generating_ipv4(self) -> None:
tests/test_pyrflx.py:    def test_ipv4_generating_ipv4_option(self) -> None:
tests/test_pyrflx.py:    # rflx-tlv-tests.adb
tests/test_pyrflx.py:    def test_tlv_parsing_tlv_data(self) -> None:
tests/test_pyrflx.py:    def test_tlv_checksum_parsing_tlv_data_zero(self) -> None:
tests/test_pyrflx.py:    def test_tlv_checksum_parsing_tlv_error(self) -> None:
tests/test_pyrflx.py:    def test_tlv_checksum_parsing_invalid_tlv_invalid_tag(self) -> None:
tests/test_pyrflx.py:    def test_tlv_generating_tlv_data(self) -> None:
tests/test_pyrflx.py:    def test_tlv_checksum_generating_tlv_error(self) -> None:
```

1. Why is `tlv_checksum` used in `tlv` tests? There are no `tlv_checksum` tests in `rflx-tlv-tests.adb`.
2. Where is `Test_Generating_TLV_Data_Zero`?
3. Why is `test_parsing_ipv4_option` not implemented?
4. Where are the parsing tests of `rflx-in_ipv4-tests.adb`?
5. Where is the comment for `rflx-in_ethernet-tests.adb`?
6. There are multiple Ethernet tests which are not correctly named! (There should be always either `generating` or `parsing` in the name.)

Such inconsistencies would make it very hard to keep these tests in sync in the future!

Reply:
@treiher @jklmnn @rssen I wonder if we could derive those "shared" tests from an abstract base class such that both worlds, SPARK and Python, are forced to implement the respective test methods. This could help to detect missing tests in the future.

Does anyone know a way to finalize a class in python such that no additional methods can be added? This would be helpful to ensure that new tests are added to the base class first.

I don't think we should implement that within this PR, but it could be worth a try.

Reply:
Fixed by f30459d. I hope that I got everything now, the "ethenet" typo was fixed in 0ba3390.

Reply:
> I wonder if we could derive those "shared" tests from an abstract base class such that both worlds, SPARK and Python, are forced to implement the respective test methods. This could help to detect missing tests in the future.

I like that idea. That would be really helpful! Maybe you could create a new issue for that.

> Does anyone know a way to finalize a class in python such that no additional methods can be added? This would be helpful to ensure that new tests are added to the base class first.

Since Python 3.8 there is a final qualifier for this purpose ([PEP 591](https://www.python.org/dev/peps/pep-0591/)). But I do not see how that would help us here?

Reply:
I created #209 and placed it into Future for now.

You're right, finalization does not help here. We need a mechanism that prevents the addition of new methods to the derived class so that you're forced to add them to the base class to implement them (and thus are also forced to add them to the other derived class).